compose: Add functionality for audio call.

sbansal1999 opened the pull request:
This PR adds the feature to add an audio call (video call with video muted by default) to the compose box.  Right now, the button to add that is only shown when the video call provider is set to `Jitsi Meet`. 

Fixes: #12207

Right now the PR does not have tests but I can add them once the design is finalized.

---

<!-- If the PR makes UI changes, always include one or more still screenshots to demonstrate your changes. If it seems helpful, add a screen capture of the new functionality as well.

Tooling tips: https://zulip.readthedocs.io/en/latest/tutorials/screenshot-and-gif-software.html
-->

**Screenshots and screen captures:**


<details>

  

<summary>Adding Audio Call to the Chat</summary>

  

[Screencast from 07-06-23 04:49:20 PM IST.webm](https://github.com/sbansal1999/zulip/assets/35286603/86d068b0-a0dd-4a58-a362-53d718756ea6)

  

</details>

  

<details>

  

<summary>Changing Video Call Provider to None</summary>

  

[Screencast from 07-06-23 05:06:07 PM IST.webm](https://github.com/sbansal1999/zulip/assets/35286603/a6ca8fa0-5d45-4eaa-bd4c-abbd5d84490e)

  

</details>

  ---

<details>
<summary>Self-review checklist</summary>

<!-- Prior to submitting a PR, follow our step-by-step guide to review your own code:
https://zulip.readthedocs.io/en/latest/contributing/code-reviewing.html#how-to-review-code -->

<!-- Once you create the PR, check off all the steps below that you have completed.
If any of these steps are not relevant or you have not completed, leave them unchecked.-->

- [x] [Self-reviewed](https://zulip.readthedocs.io/en/latest/contributing/code-reviewing.html#how-to-review-code) the changes for clarity and maintainability
      (variable names, code reuse, readability, etc.).

Communicate decisions, questions, and potential concerns.

- [ ] Explains differences from previous plans (e.g., issue description).
- [ ] Highlights technical choices and bugs encountered.
- [ ] Calls out remaining decisions and concerns.
- [ ] Automated tests verify logic where appropriate.

Individual commits are ready for review (see [commit discipline](https://zulip.readthedocs.io/en/latest/contributing/commit-discipline.html)).

- [x] Each commit is a coherent idea.
- [x] Commit message(s) explain reasoning and motivation for changes.

Completed manual review and testing of the following:

- [x] Visual appearance of the changes.
- [x] Responsiveness and internationalization.
- [x] Strings and tooltips.
- [x] End-to-end functionality of buttons, interactions and flows.
- [x] Corner cases, error conditions, and easily imagined bugs.
</details>

sbansal1999 commented:
@zulipbot add "QA needed" "area: compose"

amanagr commented:
LGTM!

timabbott commented on web/src/compose.js:
```diff
@@ -706,6 +734,27 @@ export function initialize() {
+        const show_audio_chat_button = compute_show_audio_chat_button();
+        if (!show_audio_chat_button) {
+            return;
+        }
```
Is this conditional necessary? It seems like we should be able to assume that's the case because the button was there to be clicked.

sbansal1999 replied:
This is kind of true. But this disallows the user to do a `querySelector` for the video call button and then simulate the click from the console.

timabbott replied:
OK -- that doesn't seem important as a thing to prevent throwing an exception?

Anyway, it's probably fine to have this in case we fail to live-update the button state somehow, but it seems a bit silly.

timabbott commented on web/src/compose.js:
```diff
@@ -706,6 +734,27 @@ export function initialize() {
+        }
+
+        const audio_call_link =
+            generate_jitsi_meet_video_call_link() + "#config.startWithVideoMuted=true";
```
This logic only works for Jitsi, so it shouldn't be written this way. 

Ideally, rather than having this be so parallel to the video call code path, we'd have a shared function like `generate_jitsi_meet_call_link` that takes a parameter for whether you want a video or audio call ... or ideally, a broader refactor that extracts the click handler for the video call link as a function that accepts a parameter for whether a video or audio -- that will save us a ton of time implementing the audio call feature for Zoom, for example, which likely just involves adding a similar small change to the 100-line function for asking Zoom to make a call.

sbansal1999 replied:
Agreed. That would make it easier to expand this feature to other providers as well.

sbansal1999 commented:
**Updates:**

- Extracted the whole video call link creation logic to a separate function which now is responsible for creating both audio and video call links.
- Modified a test since the logic of finding the `textarea` where the link should be inserted happens a bit later now.

timabbott commented on web/src/compose.js:
```diff
@@ -872,3 +841,95 @@ function schedule_message_to_custom_date() {
+        const video_call_link = page_params.jitsi_server_url + "/" + video_call_id;
+        if (is_audio_call) {
+            insert_audio_call_url(
+                video_call_link + "#config.startWithAudioMuted=true",
```
This disables audio, not video. I think this bug is a good reminder that you should always test code end-to-end -- if you don't do that, you're going to have bugs like this where a feature doesn't do the main thing it's supposed to...

sbansal1999 replied:
Yeah, I agree that I messed things up here.

timabbott commented on web/src/compose.js:
```diff
@@ -872,3 +841,95 @@ function schedule_message_to_custom_date() {
     });
 }
+
+function generate_and_insert_audio_or_video_call_link($target_element, is_audio_call) {
```
I refactored these parameters to accept a `$target_element` parameter here, rather than `e`.

timabbott commented:
I fixed the above issues and marked this to merge once CI passes. See #26550 and #26549 for follow-ups that will require a bit of research but hopefully not much actual work to complete this feature from an end user perspective.
